recursion-schemes-in-rust/src/lib.rs

```diff
@@ -0,0 +1,156 @@
+use std::collections::HashMap;
+
+/// The original, "naive" representation of spanned JSON values without using recursion schemes.
+mod span {
+    pub type Span = std::ops::Range<usize>;
+
+    pub struct Spanned<T> {
+        pos: Span,
+        data: T,
+    }
+
+    pub type SpannedValue = Spanned<JsonValue>;
+
+    pub enum JsonValue {
+        String(String),
+        Number(f64),
+        Pair(Box<SpannedValue>, Box<SpannedValue>),
+        Array(Vec<SpannedValue>),
+        Object(super::HashMap<String, SpannedValue>),
+    }
+}
+
+/// The functor representation.
+pub enum JsonValueF<T> {
+    String(String),
+    Number(f64),
+    Pair(T, T),
+    Array(Vec<T>),
+    Object(HashMap<String, T>),
+}
+
+/// A version of [JsonValueF] that borrows the data in the leafs. This is useful for deriving a
+/// borrowed version of `&JsonValueF<T>` without cloning the data in the leafs (versus the other
+/// contender `JsonValueF<&T>`).
+pub enum JsonValueRefF<'a, T> {
+    String(&'a String),
+    Number(&'a f64),
+    Pair(T, T),
+    Array(Vec<T>),
+    Object(HashMap<&'a str, T>),
+}
+
+pub type JsonValueRef<'a, T> = JsonValueRefF<'a, &'a T>;
+
+/// A simple JSON value from the functor representation.
+pub struct JsonValue {
+    data: JsonValueF<Box<JsonValue>>,
+}
+
+/// Same as [span::SpannedValue], but using the functor representation.
+pub struct SpannedJsonValue {
+    data: JsonValueF<Box<SpannedJsonValue>>,
+    span: std::ops::Range<u32>,
+}
+
+/// A non-owned, shared version of JSON values.
+pub struct SharedJsonValue {
+    data: JsonValueF<std::rc::Rc<JsonValue>>,
+}
+
+/// A borrowed version of JSON values.
+pub struct ArenaJsonValue<'a> {
+    data: JsonValueF<&'a JsonValue>,
+}
+
+/// Count the number of string leaves, without using recursion schemes.
+pub fn count_strings_naive(value: &JsonValue) -> u32 {
```

Comment:
Maybe 
```suggestion
pub fn count_strings_direct(value: &JsonValue) -> u32 {
```

There's nothing particularly naive about the definition.

Reply:
You also refer to the definition in the `[span]` module as "naive", which might make this confusing.